mlfinlab/feature_importance/cfi.py

```diff
@@ -0,0 +1,126 @@
+"""
+An Implementation of the Clustered Feature Importance algorithm described by
+Dr Marcos Lopez de Prado in 'Clustered Feature Importance (Presentation Slides)'
+(https://papers.ssrn.com/sol3/papers.cfm?abstract_id=3517595)
+"""
+
+#Imports
+import numpy as np
+import pandas as pd
+import mlfinlab as ml
+from sklearn.metrics import log_loss
+from scipy.cluster.hierarchy import linkage, fcluster
+from mlfinlab.codependence import variation_of_information_score
+
+def get_VI_matrix(df : pd.DataFrame):
+    '''
+    Get Variation of Information (VI) matrix.
+
+    :param df: (pd.DataFrame) of features
+    :return: (pd.DataFrame) of VI matrix
+    '''
+    VI = pd.DataFrame(index=df.columns)
+    features_cols = df.columns.values
+    for col0 in features_cols:
+        vi = []
+        for col1 in features_cols:
+            x = df[col0].values
+            y = df[col1].values
+            vi.append(variation_of_information_score(x,y,n_bins=None,normalize=True))
+        VI[col0] = vi
+    return VI
+
+def cluster_features(X : pd.DataFrame, C : int = None):
+    '''
+    Get Clusters containing features subsets
+
+    (param) X : (pd.DataFrame) of features
+    (param) C : (int) number of clusters to form.
+                      Must be less the total number of features.
+                      If None then it returns optimal number of clusters
+                      decided by the ONC Algorithm.
+
+    (return) : (array) of feature subsets
+    '''
+
+    vi_matrix =  get_VI_matrix(X)
+    link = linkage(vi_matrix, 'single')
```

Comment:
Set's also make linkage as parameter. What I've seen from my practice 'ward' linkage works much better, so let the user decide what linkage he wants to use.

Reply:
As per Dr de Prado, he mentioned only the use of a single linkage. Below I have attached a slide from the original presentation (highlighting the same). As per my understanding, he used a single linkage because it can address the issue of multicollinearity (presence of redundant features).
![cfi](https://user-images.githubusercontent.com/45006182/75976797-1e872200-5f01-11ea-974c-01472030ea56.PNG)

Reply:
Of course, de Prado also uses single-linkage algorithm in HPR. However, on practice user can deviate from presentation materials and apply other linkage functions

Reply:
Okay, I will make it a parameter and set the default to single.